ci-operator/step-registry/cucushift/upgrade/tolatest/cucushift-upgrade-tolatest-commands.sh

```diff
@@ -0,0 +1,681 @@
+    try=0
+    max_retries=2
+    response=$(https_proxy="" HTTPS_PROXY="" curl --silent --output /dev/null --write-out %"{http_code}" "https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/${algorithm}=${hash_value}/signature-1" -v)
+    while (( try < max_retries && response != 200 )); do
+        echo "Trying #${try}"
+        response=$(https_proxy="" HTTPS_PROXY="" curl --silent --output /dev/null --write-out %"{http_code}" "https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/${algorithm}=${hash_value}/signature-1" -v)
+        (( try += 1 ))
+        sleep 60
+    done
+    set +x
+    if (( response == 200 )); then
+        echo "${TARGET} is signed" && return 0
+    else
+        echo "Seem like ${TARGET} is not signed" && return 1
+    fi
+}
+
+# Check if admin ack is required before upgrade
+function admin_ack() {
+    if (( SOURCE_MINOR_VERSION == TARGET_MINOR_VERSION )) || (( SOURCE_MINOR_VERSION < 8 )); then
+        echo "Admin ack is not required in either z-stream upgrade or 4.7 and earlier" && return
+    fi
+
+    local out; out="$(oc -n openshift-config-managed get configmap admin-gates -o json | jq -r ".data")"
+    if [[ ${out} != *"ack-4.${SOURCE_MINOR_VERSION}"* ]]; then
+        echo "Admin ack not required: ${out}" && return
+    fi
+
+    echo "Require admin ack:\n ${out}"
+    local wait_time_loop_var=0 ack_data
+
+    ack_data="$(echo "${out}" | jq -r "keys[]")"
+    for ack in ${ack_data};
+    do
+        # e.g.: ack-4.12-kube-1.26-api-removals-in-4.13
+        if [[ "${ack}" == *"ack-4.${SOURCE_MINOR_VERSION}"* ]]
+        then
+            echo "Admin ack patch data is: ${ack}"
+            oc -n openshift-config patch configmap admin-acks --patch '{"data":{"'"${ack}"'": "true"}}' --type=merge
+        fi
+    done
+    echo "Admin-acks patch gets started"
+
+    echo -e "sleep 5 min wait admin-acks patch to be valid...\n"
+    while (( wait_time_loop_var < 5 )); do
+        sleep 1m
+        echo -e "wait_time_passed=${wait_time_loop_var} min.\n"
+        if ! oc adm upgrade | grep "AdminAckRequired"; then
+            echo -e "Admin-acks patch PASSED\n"
+            return 0
+        else
+            echo -e "Admin-acks patch still in processing, waiting...\n"
+        fi
+        (( wait_time_loop_var += 1 ))
+    done
+    if (( wait_time_loop_var >= 5 )); then
+        echo >&2 "Timed out waiting for admin-acks completing, exiting" && return 1
+    fi
+}
+
+# check if any of cases is enabled via ENABLE_OTA_TEST
+function check_ota_case_enabled() {
+    local case_id
+    local cases_array=("$@")
+    for case_id in "${cases_array[@]}"; do
+        # shellcheck disable=SC2076
+        if [[ " ${ENABLE_OTA_TEST} " =~ " ${case_id} " ]]; then
+            echo "${case_id} is enabled via ENABLE_OTA_TEST on this job."
+            return 0
+        fi
+    done
+    return 1
+}
+
+function set_channel(){
+    x_ver=$( echo "${DUMMY_TARGET_VERSION}" | cut -f1 -d. )
```

Comment:
On https://github.com/openshift/release/pull/49743/files#diff-c587fa8e293fdade6e68f947f6b3b4c7f27c3967a4ad98c6c7409f3a3d50ef6dR245, it is using source cluster version to determinate the target channel, while here is using dummy target payload imag version, can we keep them consistent in the two places?

Reply:
ah, because ${DUMMY_TARGET_VERSION} is ready-made which does not need `+1` on y_ver here, so i just use it directly. i'm ok to keep it consistent to use source cluster version.

Reply:
oh, just realize we could not use source cluster version +1 because it might be a z-stream upgrade with --tolatest? so  ${DUMMY_TARGET_VERSION} should be the correct one here.

As for source cluster version in prestep, it's for ocp-53907 case only, we enable it in a y-version upgrade job explicitly, so that's ok to use source cluster version, and we don't need import env OPENSHIFT_UPGRADE_RELEASE_IMAGE_OVERRIDE for it.

Reply:
> it might be a z-stream upgrade with --tolatest

Ah, I did not consider the z-stream upgrade, agree with you.